File: src/test/java/de/fraunhofer/aisec/cpg/enhancements/variable_resolution/VariableResolverCppTest.java
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.enhancements.variable_resolution;

import static org.junit.jupiter.api.Assertions.assertNotNull;

import de.fraunhofer.aisec.cpg.BaseTest;
import de.fraunhofer.aisec.cpg.TestUtils;
import de.fraunhofer.aisec.cpg.TranslationConfiguration;
import de.fraunhofer.aisec.cpg.TranslationManager;
import de.fraunhofer.aisec.cpg.graph.Node;
import de.fraunhofer.aisec.cpg.graph.declarations.FieldDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.ParamVariableDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.RecordDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.TranslationUnitDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.VariableDeclaration;
import de.fraunhofer.aisec.cpg.graph.statements.CatchClause;
import de.fraunhofer.aisec.cpg.graph.statements.CompoundStatement;
import de.fraunhofer.aisec.cpg.graph.statements.ForStatement;
import de.fraunhofer.aisec.cpg.graph.statements.IfStatement;
import de.fraunhofer.aisec.cpg.graph.statements.expressions.CallExpression;
import de.fraunhofer.aisec.cpg.graph.statements.expressions.DeclaredReferenceExpression;
import de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression;
import de.fraunhofer.aisec.cpg.graph.statements.expressions.Literal;
import de.fraunhofer.aisec.cpg.graph.statements.expressions.MemberExpression;
import de.fraunhofer.aisec.cpg.helpers.NodeComparator;
import de.fraunhofer.aisec.cpg.helpers.SubgraphWalker;
import de.fraunhofer.aisec.cpg.helpers.Util;
import java.io.File;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class VariableResolverCppTest extends BaseTest {

  private RecordDeclaration externalClass;
  private FieldDeclaration externVarName;
  private FieldDeclaration externStaticVarName;

  private RecordDeclaration outerClass;
  private FieldDeclaration outerVarName;
  private FieldDeclaration outerStaticVarName;
  private FieldDeclaration outerImpThis;

  private RecordDeclaration innerClass;
  private FieldDeclaration innerVarName;
  private FieldDeclaration innerStaticVarName;
  private FieldDeclaration innerImpThis;
  private FieldDeclaration innerImpOuter;

  private FunctionDeclaration main;

  private MethodDeclaration outerFunction1;
  private List<ForStatement> forStatements;
  private MethodDeclaration outerFunction2;
  private MethodDeclaration outerFunction3;
  private MethodDeclaration outerFunction4;
  private MethodDeclaration outerFunction5;

  private MethodDeclaration innerFunction1;
  private MethodDeclaration innerFunction2;

  private Map<String, Expression> callParamMap = new HashMap<>();

  @BeforeAll
  void initTests() throws ExecutionException, InterruptedException {
    final String topLevelPath = "src/test/resources/variables_extended/cpp/";
    List<String> fileNames = Arrays.asList("scope_variables.cpp", "external_class.cpp");
    List<File> fileLocations =
        fileNames.stream()
            .map(fileName -> new File(topLevelPath + fileName))
            .collect(Collectors.toList());
    TranslationConfiguration config =
        TranslationConfiguration.builder()
            .sourceLocations(fileLocations.toArray(new File[fileNames.size()]))
            .topLevel(new File(topLevelPath))
            .defaultPasses()
            .debugParser(true)
            .defaultLanguages()
            .failOnError(true)
            .loadIncludes(true)
            .build();

    TranslationManager analyzer = TranslationManager.builder().config(config).build();
    List<TranslationUnitDeclaration> tu = analyzer.analyze().get().getTranslationUnits();

    List<Node> nodes =
        tu.stream()
            .flatMap(tUnit -> SubgraphWalker.flattenAST(tUnit).stream())
            .collect(Collectors.toList());
    List<CallExpression> calls =
        TestUtils.findByName(Util.filterCast(nodes, CallExpression.class), "printLog");

    calls.sort(new NodeComparator());

    List<RecordDeclaration> records = Util.filterCast(nodes, RecordDeclaration.class);

    // Extract all Variable declarations and field declarations for matching
    externalClass = TestUtils.getOfTypeWithName(nodes, RecordDeclaration.class, "ExternalClass");
    externVarName =
        TestUtils.getSubnodeOfTypeWithName(externalClass, FieldDeclaration.class, "varName");
    externStaticVarName =
        TestUtils.getSubnodeOfTypeWithName(externalClass, FieldDeclaration.class, "staticVarName");
    outerClass = TestUtils.getOfTypeWithName(nodes, RecordDeclaration.class, "ScopeVariables");
    outerVarName =
        outerClass.getFields().stream()
            .filter(n -> n.getName().equals("varName"))
            .findFirst()
            .get();
    outerStaticVarName =
        outerClass.getFields().stream()
            .filter(n -> n.getName().equals("staticVarName"))
            .findFirst()
            .get();
    outerImpThis =
        outerClass.getFields().stream().filter(n -> n.getName().equals("this")).findFirst().get();

    List<RecordDeclaration> classes = Util.filterCast(nodes, RecordDeclaration.class);

    // Inner class and its fields
    innerClass =
        TestUtils.getOfTypeWithName(nodes, RecordDeclaration.class, "ScopeVariables::InnerClass");
    innerVarName =
        innerClass.getFields().stream()
            .filter(n -> n.getName().equals("varName"))
            .findFirst()
            .get();
    innerStaticVarName =
        innerClass.getFields().stream()
            .filter(n -> n.getName().equals("staticVarName"))
            .findFirst()
            .get();
    innerImpThis =
        innerClass.getFields().stream().filter(n -> n.getName().equals("this")).findFirst().get();

    main = TestUtils.getOfTypeWithName(nodes, FunctionDeclaration.class, "main");

    // Functions in the outer and inner object
    outerFunction1 =
        outerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function1"))
            .collect(Collectors.toList())
            .get(0);
    forStatements = Util.filterCast(SubgraphWalker.flattenAST(outerFunction1), ForStatement.class);
    outerFunction2 =
        outerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function2"))
            .collect(Collectors.toList())
            .get(0);
    outerFunction3 =
        outerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function3"))
            .collect(Collectors.toList())
            .get(0);
    outerFunction4 =
        outerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function4"))
            .collect(Collectors.toList())
            .get(0);
    outerFunction5 =
        outerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function5"))
            .collect(Collectors.toList())
            .get(0);
    innerFunction1 =
        innerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function1"))
            .collect(Collectors.toList())
            .get(0);
    innerFunction2 =
        innerClass.getMethods().stream()
            .filter(method -> method.getName().equals("function2"))
            .collect(Collectors.toList())
            .get(0);

    for (CallExpression call : calls) {
      Expression first = call.getArguments().get(0);
      String logId = ((Literal) first).getValue().toString();

      Expression second = call.getArguments().get(1);
      callParamMap.put(logId, second);
    }
  }

  DeclaredReferenceExpression getCallWithReference(String literal) {
    Expression exp = callParamMap.get(literal);
    if (exp instanceof DeclaredReferenceExpression) return (DeclaredReferenceExpression) exp;
    return null;
  }

  MemberExpression getCallWithMemberExpression(String literal) {
    Expression exp = callParamMap.get(literal);
    if (exp instanceof MemberExpression) return (MemberExpression) exp;
    return null;
  }

  @Test
  void testOuterVarNameAccessedImplicitThis() {
    VRUtil.assertUsageOf(callParamMap.get("func1_impl_this_varName"), outerVarName);
  }

  @Test
  void testStaticFieldAccessedImplicitly() {
    VRUtil.assertUsageOf(callParamMap.get("func1_static_staticVarName"), outerStaticVarName);
  }

  @Test
  void testVarNameOfFirstLoopAccessed() {
    DeclaredReferenceExpression asReference = getCallWithReference("func1_first_loop_varName");
    assertNotNull(asReference);
    VariableDeclaration vDeclaration =
        TestUtils.getSubnodeOfTypeWithName(
            forStatements.get(0), VariableDeclaration.class, "varName");
    VRUtil.assertUsageOf(callParamMap.get("func1_first_loop_varName"), vDeclaration);
  }

  @Test
  void testAccessLocalVarNameInNestedBlock() {
    CompoundStatement innerBlock =
        TestUtils.getSubnodeOfTypeWithName(forStatements.get(1), CompoundStatement.class, "");
    VariableDeclaration nestedDeclaration =
        TestUtils.getSubnodeOfTypeWithName(innerBlock, VariableDeclaration.class, "varName");
    VRUtil.assertUsageOf(
        callParamMap.get("func1_nested_block_shadowed_local_varName"), nestedDeclaration);
  }

  @Test
  void testVarNameOfSecondLoopAccessed() {
    VariableDeclaration vDeclaration =
        TestUtils.getSubnodeOfTypeWithName(
            forStatements.get(1), VariableDeclaration.class, "varName");
    VRUtil.assertUsageOf(callParamMap.get("func1_second_loop_varName"), vDeclaration);
  }

  @Test
  void testParamVarNameAccessed() {
    ParamVariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            outerFunction2, ParamVariableDeclaration.class, "varName");
    VRUtil.assertUsageOf(callParamMap.get("func2_param_varName"), declaration);
  }

  @Test
  void testMemberVarNameOverExplicitThis() {
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func2_this_varName"), outerImpThis, outerVarName);
  }

  @Test
  void testVarNameDeclaredInIfClause() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            TestUtils.getSubnodeOfTypeWithName(outerFunction2, IfStatement.class, Node.EMPTY_NAME),
            VariableDeclaration.class,
            "varName");
    VRUtil.assertUsageOf(callParamMap.get("func2_if_varName"), declaration);
  }

  @Test
  void testVarNameCoughtAsException() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            TestUtils.getSubnodeOfTypeWithName(outerFunction2, CatchClause.class, Node.EMPTY_NAME),
            VariableDeclaration.class,
            "varName");
    VRUtil.assertUsageOf(callParamMap.get("func2_catch_varName"), declaration);
  }

  @Test
  void testMemberAccessedOverInstance() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            outerFunction2, VariableDeclaration.class, "scopeVariables");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func2_instance_varName"), declaration, outerVarName);
  }

  @Test
  void testMemberAccessedOverInstanceAfterParamDeclaration() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            outerFunction3, VariableDeclaration.class, "scopeVariables");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func3_instance_varName"), declaration, outerVarName);
  }

  @Test
  void testAccessExternalClassMemberVarnameOverInstance() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            outerFunction3, VariableDeclaration.class, "externalClass");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func3_external_instance_varName"), declaration, externVarName);
  }

  @Test
  void testExplicitlyReferenceStaticMemberInInternalClass() {
    VRUtil.assertUsageOf(
        callParamMap.get("func4_static_staticVarName"), outerStaticVarName.getDefinition());
  }

  @Test
  void testExplicitlyReferenceStaticMemberInExternalClass() {
    VRUtil.assertUsageOf(
        callParamMap.get("func4_external_staticVarName"), externStaticVarName.getDefinition());
  }

  @Test
  void testAccessExternalMemberOverInstance() {
    VariableDeclaration externalInstance =
        TestUtils.getSubnodeOfTypeWithName(
            outerFunction4, VariableDeclaration.class, "externalClass");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func4_external_instance_varName"), externalInstance, externVarName);
  }

  @Test
  void testAccessExternalStaticMemberAfterInstanceCreation() {
    VRUtil.assertUsageOf(
        callParamMap.get("func4_second_external_staticVarName"),
        externStaticVarName.getDefinition());
  }

  @Test
  void testAccessStaticMemberThroughInstanceFirst() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(outerFunction5, VariableDeclaration.class, "first");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func5_staticVarName_throughInstance_first"),
        declaration,
        outerStaticVarName.getDefinition());
  }

  @Test
  void testAccessStaticMemberThroughInstanceSecond() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(outerFunction5, VariableDeclaration.class, "second");

    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func5_staticVarName_throughInstance_second"),
        declaration,
        outerStaticVarName.getDefinition());
  }

  @Test
  void testImplicitThisAccessOfInnerClassMember() {
    VRUtil.assertUsageOf(callParamMap.get("func1_inner_imp_this_varName"), innerVarName);
  }

  @Test
  void testAccessOfInnerClassMemberOverInstance() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(innerFunction1, VariableDeclaration.class, "inner");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func1_inner_instance_varName"), declaration, innerVarName);
  }

  @Test
  void testAccessOfOuterMemberOverInstance() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            innerFunction1, VariableDeclaration.class, "scopeVariables");

    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func1_outer_instance_varName"), declaration, outerVarName);
  }

  @Test
  void testAccessOfOuterStaticMember() {
    VRUtil.assertUsageOf(
        callParamMap.get("func1_outer_static_staticVarName"), outerStaticVarName.getDefinition());
  }

  @Test
  void testAccessOfInnerStaticMember() {
    VRUtil.assertUsageOf(
        callParamMap.get("func1_inner_static_staticVarName"), innerStaticVarName.getDefinition());
  }

  @Test
  void testAccessOfInnerClassMemberOverInstanceWithSameNamedVariable() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(innerFunction2, VariableDeclaration.class, "inner");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func2_inner_instance_varName_with_shadows"), declaration, innerVarName);
  }

  @Test
  void testAccessOfOuterMemberOverInstanceWithSameNamedVariable() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(
            innerFunction2, VariableDeclaration.class, "scopeVariables");
    VRUtil.assertUsageOfMemberAndBase(
        callParamMap.get("func2_outer_instance_varName_with_shadows"), declaration, outerVarName);
  }

  @Test
  void testAccessOfOuterStaticMembertWithSameNamedVariable() {
    VRUtil.assertUsageOf(
        callParamMap.get("func2_outer_static_staticVarName_with_shadows"),
        outerStaticVarName.getDefinition());
  }

  @Test
  void testAccessOfInnerStaticMemberWithSameNamedVariable() {
    VRUtil.assertUsageOf(
        callParamMap.get("func2_inner_static_staticVarName_with_shadows"),
        innerStaticVarName.getDefinition());
  }

  @Test
  void testLocalVariableUsedAsParameter() {
    VariableDeclaration declaration =
        TestUtils.getSubnodeOfTypeWithName(main, VariableDeclaration.class, "varName");
    VRUtil.assertUsageOf(callParamMap.get("main_local_varName"), declaration);
  }
}

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/enhancements/variable_resolution/VariableResolverCppTest.kt
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.enhancements.variable_resolution

import de.fraunhofer.aisec.cpg.BaseTest
import de.fraunhofer.aisec.cpg.TestUtils.analyze
import de.fraunhofer.aisec.cpg.TestUtils.assertUsageOf
import de.fraunhofer.aisec.cpg.TestUtils.assertUsageOfMemberAndBase
import de.fraunhofer.aisec.cpg.graph.*
import de.fraunhofer.aisec.cpg.graph.declarations.*
import de.fraunhofer.aisec.cpg.graph.statements.CatchClause
import de.fraunhofer.aisec.cpg.graph.statements.CompoundStatement
import de.fraunhofer.aisec.cpg.graph.statements.ForStatement
import de.fraunhofer.aisec.cpg.graph.statements.IfStatement
import de.fraunhofer.aisec.cpg.graph.statements.expressions.DeclaredReferenceExpression
import de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression
import de.fraunhofer.aisec.cpg.graph.statements.expressions.Literal
import java.nio.file.Path
import java.util.concurrent.ExecutionException
import kotlin.test.Test
import kotlin.test.assertNotNull
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.TestInstance

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
internal class VariableResolverCppTest : BaseTest() {
    private var externalClass: RecordDeclaration? = null
    private var externVarName: FieldDeclaration? = null
    private var externStaticVarName: FieldDeclaration? = null
    private var outerClass: RecordDeclaration? = null
    private var outerVarName: FieldDeclaration? = null
    private var outerStaticVarName: FieldDeclaration? = null
    private var function2Receiver: VariableDeclaration? = null
    private var innerClass: RecordDeclaration? = null
    private var innerVarName: FieldDeclaration? = null
    private var innerStaticVarName: FieldDeclaration? = null
    private var main: FunctionDeclaration? = null
    private var outerFunction1: MethodDeclaration? = null
    private var forStatements: List<ForStatement>? = null
    private var outerFunction2: MethodDeclaration? = null
    private var outerFunction3: MethodDeclaration? = null
    private var outerFunction4: MethodDeclaration? = null
    private var outerFunction5: MethodDeclaration? = null
    private var innerFunction1: MethodDeclaration? = null
    private var innerFunction2: MethodDeclaration? = null
    private val callParamMap: MutableMap<String, Expression> = HashMap()

    @BeforeAll
    @Throws(ExecutionException::class, InterruptedException::class)
    fun initTests() {
        val topLevel = Path.of("src/test/resources/variables_extended/cpp/")
        val files =
            listOf("scope_variables.cpp", "external_class.cpp").map {
                topLevel.resolve(it).toFile()
            }
        val result = analyze(files, topLevel, true)
        val calls = result.calls { it.name == "printLog" }
        val records = result.records
        val functions = result.functions

        // Extract all Variable declarations and field declarations for matching
        externalClass = records["ExternalClass"]
        externVarName = externalClass.fields["varName"]
        externStaticVarName = externalClass.fields["staticVarName"]
        outerClass = records["ScopeVariables"]
        outerVarName = outerClass?.byNameOrNull("varName")
        outerStaticVarName = outerClass?.byNameOrNull("staticVarName")
        function2Receiver = outerClass?.byNameOrNull<MethodDeclaration>("function2")?.receiver

        // Inner class and its fields
        innerClass = records["ScopeVariables::InnerClass"]
        innerVarName = innerClass.fields["varName"]
        innerStaticVarName = innerClass.fields["staticVarName"]
        main = functions["main"]

        // Functions in the outer and inner object
        outerFunction1 =
            outerClass?.methods?.first { method: MethodDeclaration -> method.name == "function1" }
        forStatements = outerFunction1.allChildren()
        outerFunction2 =
            outerClass?.methods?.first { method: MethodDeclaration -> method.name == "function2" }
        outerFunction3 =
            outerClass?.methods?.first { method: MethodDeclaration -> method.name == "function3" }
        outerFunction4 =
            outerClass?.methods?.first { method: MethodDeclaration -> method.name == "function4" }
        outerFunction5 =
            outerClass?.methods?.first { method: MethodDeclaration -> method.name == "function5" }
        innerFunction1 =
            innerClass?.methods?.first { method: MethodDeclaration -> method.name == "function1" }
        innerFunction2 =
            innerClass?.methods?.first { method: MethodDeclaration -> method.name == "function2" }
        for (call in calls) {
            val first = call.arguments[0]
            val logId = (first as Literal<*>).value.toString()
            val second = call.arguments[1]
            callParamMap[logId] = second
        }
    }

    @Test
    fun testOuterVarNameAccessedImplicitThis() {
        assertUsageOf(callParamMap["func1_impl_this_varName"], outerVarName)
    }

    @Test
    fun testStaticFieldAccessedImplicitly() {
        assertUsageOf(callParamMap["func1_static_staticVarName"], outerStaticVarName)
    }

    @Test
    fun testVarNameOfFirstLoopAccessed() {
        val asReference = callParamMap["func1_first_loop_varName"] as? DeclaredReferenceExpression
        assertNotNull(asReference)
        val vDeclaration = forStatements?.first().variables["varName"]
        assertUsageOf(callParamMap["func1_first_loop_varName"], vDeclaration)
    }

    @Test
    fun testAccessLocalVarNameInNestedBlock() {
        val innerBlock = forStatements?.get(1).allChildren<CompoundStatement>()[""]
        val nestedDeclaration = innerBlock.variables["varName"]
        assertUsageOf(callParamMap["func1_nested_block_shadowed_local_varName"], nestedDeclaration)
    }

    @Test
    fun testVarNameOfSecondLoopAccessed() {
        val vDeclaration = forStatements?.get(1).variables["varName"]
        assertUsageOf(callParamMap["func1_second_loop_varName"], vDeclaration)
    }

    @Test
    fun testParamVarNameAccessed() {
        val declaration = outerFunction2?.parameters["varName"]
        assertUsageOf(callParamMap["func2_param_varName"], declaration)
    }

    @Test
    fun testMemberVarNameOverExplicitThis() {
        assertUsageOfMemberAndBase(
            callParamMap["func2_this_varName"],
            function2Receiver,
            outerVarName
        )
    }

    @Test
    fun testVarNameDeclaredInIfClause() {
        val declaration =
            outerFunction2.allChildren<IfStatement>()[Node.EMPTY_NAME].variables["varName"]
        assertUsageOf(callParamMap["func2_if_varName"], declaration)
    }

    @Test
    fun testVarNameCoughtAsException() {
        val declaration = outerFunction2.allChildren<CatchClause>()[""].variables["varName"]
        assertUsageOf(callParamMap["func2_catch_varName"], declaration)
    }

    @Test
    fun testMemberAccessedOverInstance() {
        val declaration = outerFunction2.variables["scopeVariables"]
        assertUsageOfMemberAndBase(
            callParamMap["func2_instance_varName"],
            declaration,
            outerVarName
        )
    }

    @Test
    fun testMemberAccessedOverInstanceAfterParamDeclaration() {
        val declaration = outerFunction3.variables["scopeVariables"]
        assertUsageOfMemberAndBase(
            callParamMap["func3_instance_varName"],
            declaration,
            outerVarName
        )
    }

    @Test
    fun testAccessExternalClassMemberVarnameOverInstance() {
        val declaration = outerFunction3.variables["externalClass"]
        assertUsageOfMemberAndBase(
            callParamMap["func3_external_instance_varName"],
            declaration,
            externVarName
        )
    }

    @Test
    fun testExplicitlyReferenceStaticMemberInInternalClass() {
        assertUsageOf(callParamMap["func4_static_staticVarName"], outerStaticVarName?.definition)
    }

    @Test
    fun testExplicitlyReferenceStaticMemberInExternalClass() {
        assertUsageOf(callParamMap["func4_external_staticVarName"], externStaticVarName?.definition)
    }

    @Test
    fun testAccessExternalMemberOverInstance() {
        val externalInstance = outerFunction4.variables["externalClass"]
        assertUsageOfMemberAndBase(
            callParamMap["func4_external_instance_varName"],
            externalInstance,
            externVarName
        )
    }

    @Test
    fun testAccessExternalStaticMemberAfterInstanceCreation() {
        assertUsageOf(
            callParamMap["func4_second_external_staticVarName"],
            externStaticVarName?.definition
        )
    }

    @Test
    fun testAccessStaticMemberThroughInstanceFirst() {
        val declaration = outerFunction5.variables["first"]
        assertUsageOfMemberAndBase(
            callParamMap["func5_staticVarName_throughInstance_first"],
            declaration,
            outerStaticVarName?.definition
        )
    }

    @Test
    fun testAccessStaticMemberThroughInstanceSecond() {
        val declaration = outerFunction5.variables["second"]
        assertUsageOfMemberAndBase(
            callParamMap["func5_staticVarName_throughInstance_second"],
            declaration,
            outerStaticVarName?.definition
        )
    }

    @Test
    fun testImplicitThisAccessOfInnerClassMember() {
        assertUsageOf(callParamMap["func1_inner_imp_this_varName"], innerVarName)
    }

    @Test
    fun testAccessOfInnerClassMemberOverInstance() {
        val declaration = innerFunction1.variables["inner"]
        assertUsageOfMemberAndBase(
            callParamMap["func1_inner_instance_varName"],
            declaration,
            innerVarName
        )
    }

    @Test
    fun testAccessOfOuterMemberOverInstance() {
        val declaration = innerFunction1.variables["scopeVariables"]
        assertUsageOfMemberAndBase(
            callParamMap["func1_outer_instance_varName"],
            declaration,
            outerVarName
        )
    }

    @Test
    fun testAccessOfOuterStaticMember() {
        assertUsageOf(
            callParamMap["func1_outer_static_staticVarName"],
            outerStaticVarName?.definition
        )
    }

    @Test
    fun testAccessOfInnerStaticMember() {
        assertUsageOf(
            callParamMap["func1_inner_static_staticVarName"],
            innerStaticVarName?.definition
        )
    }

    @Test
    fun testAccessOfInnerClassMemberOverInstanceWithSameNamedVariable() {
        val declaration = innerFunction2.variables["inner"]
        assertUsageOfMemberAndBase(
            callParamMap["func2_inner_instance_varName_with_shadows"],
            declaration,
            innerVarName
        )
    }

    @Test
    fun testAccessOfOuterMemberOverInstanceWithSameNamedVariable() {
        val declaration = innerFunction2.variables["scopeVariables"]
        assertUsageOfMemberAndBase(
            callParamMap["func2_outer_instance_varName_with_shadows"],
            declaration,
            outerVarName
        )
    }

    @Test
    fun testAccessOfOuterStaticMemberWithSameNamedVariable() {
        assertUsageOf(
            callParamMap["func2_outer_static_staticVarName_with_shadows"],
            outerStaticVarName?.definition
        )
    }

    @Test
    fun testAccessOfInnerStaticMemberWithSameNamedVariable() {
        assertUsageOf(
            callParamMap["func2_inner_static_staticVarName_with_shadows"],
            innerStaticVarName?.definition
        )
    }

    @Test
    fun testLocalVariableUsedAsParameter() {
        val declaration = main.variables["varName"]
        assertUsageOf(callParamMap["main_local_varName"], declaration)
    }
}
